src/index.ts

```diff
@@ -1,40 +1,78 @@
-import { Request } from "apollo-server-env";
 import {
   ApolloServerPlugin,
+  BaseContext,
+  GraphQLFieldResolverParams,
   GraphQLRequestContext,
   GraphQLRequestListener,
-  GraphQLRequestExecutionListener,
-  GraphQLFieldResolverParams,
   GraphQLRequestContextDidEncounterErrors,
-} from "apollo-server-plugin-base";
-import { DocumentNode, GraphQLResolveInfo } from "graphql";
+  GraphQLRequestExecutionListener,
+} from "@apollo/server";
+import { GraphQLResolveInfo } from "graphql";
 import { FORMAT_HTTP_HEADERS, Span, Tracer } from "opentracing";
-import {
-  addContextHelpers,
-  SpanContext,
-  requestIsInstanceContextRequest,
-} from "./context";
+import { addContextHelpers, SpanContext } from "./context";
 
 export { SpanContext, addContextHelpers };
 
+// used for sneakily attatching our SpanContext object to the application's GraphQL context object
+const spanContextSymbol: unique symbol = Symbol.for(
+  "apolloOpenTracingSpanContext"
+);
+
+/**
+ * The interface for the application's context object
+ */
+export interface ContextForApolloOpenTracing extends BaseContext {
+  // Initially empty - we will add this to the context object.
+  [spanContextSymbol]?: SpanContext;
+  // May be specified by users when creating their context object
+  requestSpan?: Span;
+}
+
+/**
+ * Our version of the application's context object (with the span context provided)
+ */
+export type WithRequiredSpanContext<T> = T & {
+  [spanContextSymbol]: SpanContext;
+};
+
+/**
+ * Does a GraphQL context object already contain a SpanContext property?
+ */
+export function hasSpanContext<Context extends object>(
+  contextValue: Context
+): contextValue is WithRequiredSpanContext<Context> {
+  return spanContextSymbol in contextValue;
+}
+
+/**
+ * Attaches a SpanContext property to a GraphQL context object.
+ * (Has a side effect by mutating contextValue.)
+ */
+export function addSpanContext<Context extends ContextForApolloOpenTracing>(
+  contextValue: Context,
+  spanContext: SpanContext
+): asserts contextValue is WithRequiredSpanContext<Context> {
+  contextValue[spanContextSymbol] = spanContext;
+}
+
 const alwaysTrue = () => true;
 const emptyFunction = () => {};
 
-export interface InitOptions<TContext> {
+export interface InitOptions<TContext extends BaseContext> {
   server?: Tracer;
   local?: Tracer;
   onFieldResolveFinish?: (error: Error | null, result: any, span: Span) => void;
   onFieldResolve?: (
     source: any,
     args: { [argName: string]: any },
-    context: SpanContext,
+    context: TContext,
```

Comment:
why did the lifecycle methods need to know about the SpanContext? (supposed to be private?)